hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -0,0 +1,381 @@
+    private final ServerName master;
+    private final RpcCall<RESP> rpcCall;
+    private final Function<RESP, Boolean> isValidResp;
+    private final String debugStr;
+
+    MasterRpc(BatchRpcCtx<RESP> rpcCtx, ServerName master, RpcCall<RESP> rpcCall,
+       Function<RESP, Boolean> isValidResp, String debugStr) {
+      this.rpcCtx = rpcCtx;
+      this.master = master;
+      this.rpcCall = rpcCall;
+      this.isValidResp = isValidResp;
+      this.debugStr = debugStr;
+    }
+
+    @Override
+    public void run() {
+      try {
+        RESP resp = rpcCall.doRpc(getMasterStub(master));
+        if (isValidResp.apply(resp)) {
+          // Valid result, set if not set by other threads.
+          rpcCtx.setResultIfNotSet(resp);
+          return;
+        }
+      } catch (Exception e) {
+        LOG.warn("Error calling {} on master {}. Trying other masters.", debugStr, master, e);
+      }
+      rpcCtx.incrementFailedRPCs();
+    }
+  }
+
+  MasterRegistry(Configuration conf) {
+    masterServers = new ArrayList<>();
+    requestFanOut =
+        conf.getInt(MASTER_REGISTRY_NUM_HEDGED_REQS_KEY, MASTER_REGISTRY_NUM_HEDGED_REQS_DEFAULT);
+    Preconditions.checkArgument(requestFanOut >= 1);
+    if (requestFanOut > 1) {
+      masterRpcPool = Executors.newFixedThreadPool(requestFanOut,
+          new ThreadFactoryBuilder().setDaemon(true).setNameFormat("MasterRegistryRPC-%d").build());
+    }
+    parseMasterAddrs(conf);
+    rpcTimeoutNs = (int) Math.min(Integer.MAX_VALUE,
+        TimeUnit.MILLISECONDS.toNanos(conf.getLong(HBASE_RPC_TIMEOUT_KEY,
+        DEFAULT_HBASE_RPC_TIMEOUT)));
+    // TODO(HBASE-23330): Fix clients using cluster ID based token auth.
+    rpcClient = RpcClientFactory.createClient(conf, HConstants.CLUSTER_ID_DEFAULT);
+    rpcControllerFactory = RpcControllerFactory.instantiate(conf);
+  }
+
+  /**
+   * Parses the list of master addresses from the provided configuration.
+   * @param conf Configuration to parse from.
+   */
+  private void parseMasterAddrs(Configuration conf) {
+    String configuredMasters = conf.get(MASTER_ADDRS_KEY, MASTER_ADDRS_DEFAULT);
+    for (String masterAddr: configuredMasters.split(",")) {
+      masterServers.add(ServerName.valueOf(masterAddr, ServerName.NON_STARTCODE));
+    }
+    // (Pseudo) Randomized so that not all clients hot spot the same set of masters.
+    Collections.shuffle(masterServers);
+    Preconditions.checkArgument(!masterServers.isEmpty());
+  }
+
+  /**
+   * Makes a given RPC to master servers.
+   * @param rpcCall Call to make.
+   * @param debug String used for debug logging the RPC details.
+   * @param <RESP> Response type for the RPC.
+   * @param isvalidResp Used to verify if the response returned from RPC is valid.
+   * @return Optional response from the RPCs to parsed masters.
+   */
+  @VisibleForTesting
+  <RESP> Optional<RESP> doRPCs(RpcCall<RESP> rpcCall,
+      Function<RESP, Boolean> isvalidResp, String debug) {
+    if (requestFanOut == 1) {
```

Comment:
I was expecting for the first cut we'd do just one request to a random host on the list at a time, and retry with another random choice. (So above randomization of list is good and important.) This is what the zookeeper client does now is no different from current state of play. 

Hedged reading is ahead of the game.

Good that it is off by default, though.

Also, it's nice that fan out factor is configurable, but I would want an adaptive policy by default. Only if single requests are failing at some threshold of unacceptable probability (i suppose controlled by a config param) would you want to start loading up more than one per request in trade for faster response, hopefully, on average.

Reply:
Let's generalize this and apply it to and reuse the existing RPC controller, RPC client, Callable, Call, Caller hierarchy.